noobaa/noobaa-operator
# Use Service Account for KMS authentication

Includes #796
Access KMS using libopenstorage/secrets library

1. Extract KMS utils into a separate file
2. Add KMS status condition mechanism
- Types
```golang
       ConditionTypeKMSStatus conditionsv1.ConditionType = "KMS-Status"
       ConditionTypeKMSType   conditionsv1.ConditionType = "KMS-Type"
```
- Statuses
```golang
	// External KMS initialized
	ConditionKMSInit        corev1.ConditionStatus = "Init"

	// The root key was synchronized from external KMS
	ConditionKMSSync        corev1.ConditionStatus = "Sync"

	// Invalid external KMS definition
	ConditionKMSInvalid     corev1.ConditionStatus = "Invalid"

	// Error reading secret from external KMS
	ConditionKMSErrorRead   corev1.ConditionStatus = "ErrorRead"

	// Error writing initial root key to eternal KMS
	ConditionKMSErrorWrite  corev1.ConditionStatus = "ErrorWrite"
```
3. Add KMS integration tests


4. KMS: TLS and ServiceAccount auth
    
- Extract KSM TLS (ServiceAccount) tests. Use rook vault deploy
procedure.
- Add Vault V2 test
- Address review comments
- Update libopenstorage/secret mod for Vault V2 key destroy
   
5. Use libopenstorage/secrets both for k8s and vault
- Same reconcile logic for k8s/vault, see [ReconcileRootSecret()](https://github.com/baum/noobaa-operator/blob/8b71a975d2d0d8d4e3bc846c5efff3322c31c0de/pkg/system/phase2_creating.go#L736-L774) and [deleteRootSecret()](https://github.com/baum/noobaa-operator/blob/19a818a75eed0d3d77cf7bb583da9f4e832daa6b/pkg/system/reconciler.go#L397-L412)
- Extract KMSSecret and KMSDriver interfaces
- Implement KMSVault and KMSK8S drivers

Signed-off-by: Alexander Indenbaum <aindenba@redhat.com>
